ruma-client-api/src/r0/session/get_login_types.rs

```diff
@@ -40,51 +48,345 @@ impl Response {
+    pub fn login_type(&self) -> &str {
+        match self {
+            Self::Password(_) => "m.login.password",
+            Self::Token(_) => "m.login.token",
+            Self::Sso(_) => "m.login.sso",
+            Self::_Custom(c) => &c.type_,
+        }
+    }
+
+    /// Returns the associated data.
+    ///
+    /// Prefer to use the public variants of `LoginType` where possible; this method is meant to
+    /// be used for unsupported login types only.
+    pub fn data(&self) -> Cow<'_, JsonObject> {
+        fn serialize<T: Serialize>(obj: &T) -> JsonObject {
+            match serde_json::to_value(obj).expect("login type serialization to succeed") {
+                JsonValue::Object(obj) => obj,
+                _ => panic!("all login types must serialize to objects"),
+            }
+        }
+
+        match self {
+            Self::Password(d) => Cow::Owned(serialize(d)),
+            Self::Token(d) => Cow::Owned(serialize(d)),
+            Self::Sso(d) => Cow::Owned(serialize(d)),
+            Self::_Custom(c) => Cow::Borrowed(&c.data),
+        }
+    }
+}
+
+/// The payload for password login.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "type", rename = "m.login.password")]
+pub struct PasswordLoginType {}
+
+impl PasswordLoginType {
+    /// Creates a new `PasswordLoginType`.
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for PasswordLoginType {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// The payload for token-based login.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "type", rename = "m.login.token")]
+pub struct TokenLoginType {}
+
+impl TokenLoginType {
+    /// Creates a new `PasswordLoginType`.
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Default for TokenLoginType {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// The payload for SSO login.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "type", rename = "m.login.sso")]
+pub struct SsoLoginType {
+    /// The identity provider choices.
+    ///
+    /// This uses the unstable prefix in
+    /// [MSC2858](https://github.com/matrix-org/matrix-doc/pull/2858).
+    #[cfg(feature = "unstable-pre-spec")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+    #[serde(rename = "org.matrix.msc2858.identity_providers")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub identity_providers: Option<Vec<IdentityProvider>>,
+}
+
+impl SsoLoginType {
+    /// Creates a new `PasswordLoginType`.
+    pub fn new() -> Self {
+        Self {
+            #[cfg(feature = "unstable-pre-spec")]
+            identity_providers: None,
+        }
```

Comment:
```suggestion
        Self::default()
```

plus derived default is less code for the same thing :)